tests/test_types.py

```diff
@@ -0,0 +1,81 @@
+from dataclasses import dataclass
+from sys import version_info
+from typing import Optional, Union
+
+from pytest import mark, raises
+
+from pydargs import parse
+
+
+class TestUnion:
+    @dataclass
+    class Config:
+        a: int
+        b: Optional[int] = None
+        c: Union[int, str] = "b"
+        d: Union[None, int] = None
+        e: Optional[Union[int, float]] = None
+        f: float = 2.0
+
+    def test_default(self):
+        config = parse(self.Config, ["--a", "1"])
+        assert config.a == 1
+        assert config.b is None
+        assert config.c == "b"
+        assert config.d is None
+        assert config.e is None
+
+    def test_parse_optional(self):
+        config = parse(self.Config, ["--a", "1", "--b", "2"])
+        assert config.b == 2
+
+    def test_parse_optional_invalid(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, ["--a", "1", "--b", "2.0"])
+        captured = capsys.readouterr()
+        assert "argument --b: invalid functools.partial" in captured.err
+
+    @mark.parametrize("value, result", [("1.0", "1.0"), ("11", 11)])
+    def test_parse_union(self, value: str, result: Union[int, str]):
+        config = parse(self.Config, ["--a", "1", "--c", value])  # type: ignore
+        assert config.c == result
+
+    def test_parse_inverted_optional(self):
+        config = parse(self.Config, ["--a", "1", "--d", "2"])
+        assert config.d == 2
+
+    @mark.parametrize("value, result", [("1.0", 1.0), ("11", 11)])
+    def test_parse_optional_union(self, value: str, result: Union[int, str]):
+        config = parse(self.Config, ["--a", "1", "--e", value])  # type: ignore
+        assert config.e == result
+
+
+if version_info >= (3, 10):
+
+    class TestUnionType:
+        @dataclass
+        class Config:
+            a: int
+            b: int | None = None
+            c: int | str = "b"
+
+        def test_default(self):
+            config = parse(self.Config, ["--a", "1"])
+            assert config.a == 1
+            assert config.b is None
+            assert config.c == "b"
+
+        def test_parse_optional(self):
+            config = parse(self.Config, ["--a", "1", "--b", "2"])
+            assert config.b == 2
```

Comment:
related to my comment above, can you add the test case for c as well, since that has a default value of type str, this would then be cast to int due to the order of the typehint.

In principle it shouldn't matter of course as people indicate this stuff themselves, so it should work with both.. so we can also leave this out if you want

Reply:
Isn't that lines 78-81?

Reply:
Sort of, the first case can not be cast to an integer I guess? 

That's why it stays str, but it's still a number, so I added another test case that makes this a little more obvious

Reply:
I was just surprised the default has str type but then the order of the typehints swaps this around, I would expect the first priority to be to match the type of the default value

Reply:
`int("1.0")` will indeed not parse to an integer.

I don't think the type of the default should have anything to do with how we parse - especially in the case of `a_field: int | None = None` where you would never want to parse to `None`.

Reply:
Good point